django/db/models/functions/text.py

```diff
@@ -73,6 +73,23 @@ class ConcatPair(Func):
 
     function = "CONCAT"
 
+    def _resolve_output_field(self):
+        text_field = None
+        max_length = 0
+        for field in self.get_source_fields():
+            if isinstance(field, TextField):
+                text_field = field
+            elif (
+                isinstance(field, CharField)
+                and (field_max_length := field.max_length) is not None
```

Comment:
We could arguably _tolerate_ `CharField(max_length=None)` by treating it just like `TextField` but I figured it'd be better to play it safe at first. In practice there are a lot of core functions that would benefit from a more precise `output_field` (e.g. `MD5.output_field = CharField(32)`)